src/cfg/cfg-traversal.h

```diff
@@ -210,41 +219,89 @@ struct CFGWalker : public ControlFlowWalker<SubType, VisitorType> {
     // unless the call is within a try-catch, because the CFG will have too many
     // blocks that way, and if an exception is thrown, the function will be
     // exited anyway.
-    if (!self->catchStack.empty()) {
+    if (!self->unwindCatchStack.empty()) {
       auto* last = self->currBasicBlock;
       self->startBasicBlock();
       self->link(last, self->currBasicBlock);    // exception not thrown
-      self->link(last, self->catchStack.back()); // exception thrown
+      // Exception thrown. Create a link to each catch within the innermost try.
+      for (auto* block : self->unwindCatchStack.back()) {
+        self->link(last, block);
+      }
+      // If the innermost try does not have a catch_all clause, an exception
+      // thrown can be caught by any of its outer catch block. And if that outer
+      // try-catch also does not have a catch_all, this continues until we
+      // encounter try-catch_all. Create a link to all those possible catch
+      // unwind destinations.
+      for (int i = self->unwindCatchStack.size() - 1; i > 0; i--) {
+        if (self->unwindExprStack[i]->template cast<Try>()->hasCatchAll()) {
+          break;
+        }
+        for (auto* block : self->unwindCatchStack[i - 1]) {
+          self->link(last, block);
+        }
+      }
     }
   }
 
   static void doStartTry(SubType* self, Expression** currp) {
+    auto* curr = (*currp)->cast<Try>();
     auto* last = self->currBasicBlock;
-    self->startBasicBlock(); // catch body's first block
-    self->catchStack.push_back(self->currBasicBlock);
+    self->unwindCatchStack.emplace_back();
+    self->unwindExprStack.push_back(curr);
+    for (Index i = 0; i < curr->catchBodies.size(); i++) {
+      self->startBasicBlock(); // catch body's first block
+      self->unwindCatchStack.back().push_back(self->currBasicBlock);
+    }
     self->currBasicBlock = last; // reset to the current block
   }
 
-  static void doStartCatch(SubType* self, Expression** currp) {
+  static void doStartCatches(SubType* self, Expression** currp) {
     self->tryStack.push_back(self->currBasicBlock); // last block of try body
-    self->currBasicBlock = self->catchStack.back();
-    self->catchStack.pop_back();
+    self->processCatchStack.push_back(self->unwindCatchStack.back());
+    self->unwindCatchStack.pop_back();
+    self->unwindExprStack.pop_back();
+  }
+
+  static void doStartCatch(SubType* self, Expression** currp, Index i) {
+    self->currBasicBlock = self->processCatchStack.back()[i];
+  }
+
+  static void doEndCatch(SubType* self, Expression** currp, Index i) {
+    self->processCatchStack.back()[i] = self->currBasicBlock;
   }
```

Comment:
I don't follow what these two methods do. The start assigns the current basic block, and the end does the reverse assignment - is it not redundant? If not, what is it writing over?

Reply:
When there's N catches, `doStartTry` creates N BBs, each of which will serve as the start BB for each catch, and store them in `unwindCatchBlock`, so that any throwing instructions within the try body can create a link to them.

In `doStartCatches`, we move those N catch start BBs to `processCatchBlock`. We need this vector because when we enter each catch in `doStartCatch`, we should be able to retrieve the start BB created for that catch. But we can't leave those catch start BBs in `unwindCatchBlock` anymore because that stack is for processing unwinding, the process I described above (= any throwing instruction needs a link to every catch start BB in `unwindCatchBlock.back()`).

In `doEndCatch`, we save the end BB of each catch to `processCatchStack`. They are used in `doEndTry` that we create a link from each of those end catch BB to the continuation BB after the try-catch. Reusing `processCatchStack` is just to save some space, and we can use a separate vector for this; let me know if you think this is confusing and it'd be better to use separate vectors. 

I'm not sure if I described this well; it was hard to describe in the comments too. If you have any suggestions to improve the comments (or the code) I'd appreciate that.

Reply:
Thanks, I think I'm starting to get it.

Separate vectors is another option. Not sure it's worth the overhead, as I'm not sure how many catches we expect to see in practice...?

If we keep the same vector as now, I think comments in these two methods could be helpful, like
```
// Get the block that starts this catch

// We are done with this catch; set the block that ends it
```
Even though they don't add much information, they increase the chance of it making sense when reading just those two methods, I think.

Reply:
Done